### ISSUE TITLE
Deadlock in uwsgi_python_master_fixup

### ISSUE THREAD

**Issue by mafanasev** [unbit/uwsgi#1149]
Consider example:
```
import bz2
import os

from threading import Thread, Lock
from time import sleep

from flask import Flask
from pymongo import MongoClient

app = Flask(__name__)


def target():
    while True:
        # Any of these cases are leading to deadlock
        client = MongoClient()
        # bz2.compress('*' * 100000)
        # Lock()
        sleep(0.1)

for i in range(20):
    thread = Thread(target=target)
    thread.setDaemon(True)
    thread.start()


@app.route('/')
def index():
    return 'OK'
```

```
uwsgi --plugins python --file app.py --master --logto uwsgi.log --processes 50 --http-socket 0.0.0.0:8080 --callable app --enable-threads --py-call-osafterfork
```

I found that some workers (about 0-10 of 50) are in deadlock in `uwsgi_python_master_fixup`.
```
$ for p in $(pgrep -f uwsgi); do gdb --batch -p $p -x bt; done | grep sem_wait
0x00007f88a0826fd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
0x00007f88a0826fd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
0x00007f88a0826fd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
0x00007f88a0826fd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
0x00007f88a0826fd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
0x00007f88a0826fd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
0x00007f88a0826fd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
0x00007f88a0826fd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
```

More info on stack:
```
$ gdb -p 10915
$ (gdb) bt
#0  0x00007f88a0826fd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f889df88e28 in PyThread_acquire_lock () from /usr/lib/libpython2.7.so.1.0
#2  0x00007f889e3136d9 in gil_real_get () from /usr/lib/uwsgi/plugins/python_plugin.so
#3  0x00007f889e30bb37 in uwsgi_python_master_fixup () from /usr/lib/uwsgi/plugins/python_plugin.so
#4  0x000000000042c54e in uwsgi_respawn_worker ()
#5  0x0000000000465806 in uwsgi_start ()
#6  0x0000000000466fd5 in uwsgi_setup ()
#7  0x00000000004167d9 in main ()
```

It seems to me that uWSGI In `uwsgi_python_master_fixup` assumes that GIL is released and tries to acquire it. But somehow it is already acquired. 

log is also available:
```
*** Starting uWSGI 2.0.6-debian (64bit) on [Thu Jan  7 00:03:19 2016] ***
compiled with version: 4.6.4 on 22 January 2015 14:10:28
os: Linux-3.2.0-68-virtual #102-Ubuntu SMP Tue Aug 12 22:14:39 UTC 2014
machine: x86_64
clock source: unix
pcre jit disabled
detected number of CPU cores: 1
current working directory: /home/user/app
detected binary path: /usr/bin/uwsgi-core
your processes number limit is 15964
your memory page size is 4096 bytes
detected max file descriptor number: 1024
lock engine: pthread robust mutexes
thunder lock: disabled (you can enable it with --thunder-lock)
uwsgi socket 0 bound to TCP address 0.0.0.0:8080 fd 4
Python version: 2.7.3 (default, Jun 22 2015, 19:44:33)  [GCC 4.6.3]
Python main interpreter initialized at 0x1d750d0
python threads support enabled
your server socket listen backlog is limited to 100 connections
your mercy for graceful operations on workers is 60 seconds
mapped 3711168 bytes (3624 KB) for 50 cores
*** Operational MODE: preforking ***
WSGI app 0 (mountpoint='') ready in 0 seconds on interpreter 0x1d750d0 pid: 10865 (default app)
*** uWSGI is running in multiple interpreter mode ***
spawned uWSGI master process (pid: 10865)
spawned uWSGI worker 1 (pid: 10909, cores: 1)
spawned uWSGI worker 2 (pid: 10910, cores: 1)
spawned uWSGI worker 3 (pid: 10911, cores: 1)
spawned uWSGI worker 4 (pid: 10912, cores: 1)
spawned uWSGI worker 5 (pid: 10913, cores: 1)
spawned uWSGI worker 6 (pid: 10914, cores: 1)
spawned uWSGI worker 7 (pid: 10915, cores: 1)
spawned uWSGI worker 8 (pid: 10916, cores: 1)
spawned uWSGI worker 9 (pid: 10917, cores: 1)
spawned uWSGI worker 10 (pid: 10918, cores: 1)
spawned uWSGI worker 11 (pid: 10919, cores: 1)
spawned uWSGI worker 12 (pid: 10921, cores: 1)
spawned uWSGI worker 13 (pid: 10922, cores: 1)
spawned uWSGI worker 14 (pid: 10923, cores: 1)
spawned uWSGI worker 15 (pid: 10924, cores: 1)
spawned uWSGI worker 16 (pid: 10925, cores: 1)
spawned uWSGI worker 17 (pid: 10926, cores: 1)
spawned uWSGI worker 18 (pid: 10927, cores: 1)
spawned uWSGI worker 19 (pid: 10928, cores: 1)
spawned uWSGI worker 20 (pid: 10929, cores: 1)
spawned uWSGI worker 21 (pid: 10930, cores: 1)
spawned uWSGI worker 22 (pid: 10931, cores: 1)
spawned uWSGI worker 23 (pid: 10932, cores: 1)
spawned uWSGI worker 24 (pid: 10933, cores: 1)
spawned uWSGI worker 25 (pid: 10934, cores: 1)
spawned uWSGI worker 26 (pid: 10935, cores: 1)
spawned uWSGI worker 27 (pid: 10936, cores: 1)
spawned uWSGI worker 28 (pid: 10937, cores: 1)
spawned uWSGI worker 29 (pid: 10938, cores: 1)
spawned uWSGI worker 30 (pid: 10939, cores: 1)
spawned uWSGI worker 31 (pid: 10940, cores: 1)
spawned uWSGI worker 32 (pid: 10941, cores: 1)
spawned uWSGI worker 33 (pid: 10942, cores: 1)
spawned uWSGI worker 34 (pid: 10943, cores: 1)
spawned uWSGI worker 35 (pid: 10944, cores: 1)
spawned uWSGI worker 36 (pid: 10945, cores: 1)
spawned uWSGI worker 37 (pid: 10946, cores: 1)
spawned uWSGI worker 38 (pid: 10947, cores: 1)
spawned uWSGI worker 39 (pid: 10948, cores: 1)
spawned uWSGI worker 40 (pid: 10949, cores: 1)
spawned uWSGI worker 41 (pid: 10950, cores: 1)
spawned uWSGI worker 42 (pid: 10951, cores: 1)
spawned uWSGI worker 43 (pid: 10971, cores: 1)
spawned uWSGI worker 44 (pid: 10972, cores: 1)
spawned uWSGI worker 45 (pid: 10973, cores: 1)
spawned uWSGI worker 46 (pid: 10974, cores: 1)
spawned uWSGI worker 47 (pid: 10975, cores: 1)
spawned uWSGI worker 48 (pid: 10976, cores: 1)
spawned uWSGI worker 49 (pid: 10977, cores: 1)
spawned uWSGI worker 50 (pid: 10978, cores: 1)
SIGINT/SIGQUIT received...killing workers...
worker 34 buried after 1 seconds
worker 1 buried after 2 seconds
worker 2 buried after 2 seconds
worker 3 buried after 2 seconds
worker 4 buried after 2 seconds
worker 5 buried after 2 seconds
worker 6 buried after 2 seconds
worker 15 buried after 2 seconds
worker 16 buried after 2 seconds
worker 17 buried after 2 seconds
worker 18 buried after 2 seconds
worker 19 buried after 2 seconds
worker 20 buried after 2 seconds
worker 21 buried after 2 seconds
worker 22 buried after 2 seconds
worker 23 buried after 2 seconds
worker 24 buried after 2 seconds
worker 25 buried after 2 seconds
worker 26 buried after 2 seconds
worker 27 buried after 2 seconds
worker 28 buried after 2 seconds
worker 29 buried after 2 seconds
worker 30 buried after 2 seconds
worker 31 buried after 2 seconds
worker 32 buried after 2 seconds
worker 33 buried after 2 seconds
worker 35 buried after 2 seconds
worker 36 buried after 2 seconds
worker 37 buried after 2 seconds
worker 38 buried after 2 seconds
worker 39 buried after 2 seconds
worker 40 buried after 2 seconds
worker 41 buried after 2 seconds
worker 42 buried after 2 seconds
worker 43 buried after 2 seconds
worker 44 buried after 2 seconds
worker 45 buried after 2 seconds
worker 46 buried after 2 seconds
worker 47 buried after 2 seconds
worker 48 buried after 2 seconds
worker 49 buried after 2 seconds
worker 50 buried after 2 seconds
Thu Jan  7 00:09:11 2016 - worker 7 (pid: 10915) is taking too much time to die...NO MERCY !!!
Thu Jan  7 00:09:11 2016 - worker 8 (pid: 10916) is taking too much time to die...NO MERCY !!!
Thu Jan  7 00:09:11 2016 - worker 9 (pid: 10917) is taking too much time to die...NO MERCY !!!
Thu Jan  7 00:09:11 2016 - worker 10 (pid: 10918) is taking too much time to die...NO MERCY !!!
Thu Jan  7 00:09:11 2016 - worker 11 (pid: 10919) is taking too much time to die...NO MERCY !!!
Thu Jan  7 00:09:11 2016 - worker 12 (pid: 10921) is taking too much time to die...NO MERCY !!!
Thu Jan  7 00:09:11 2016 - worker 13 (pid: 10922) is taking too much time to die...NO MERCY !!!
Thu Jan  7 00:09:11 2016 - worker 14 (pid: 10923) is taking too much time to die...NO MERCY !!!
worker 7 buried after 1 seconds
worker 8 buried after 1 seconds
worker 9 buried after 1 seconds
worker 10 buried after 1 seconds
worker 11 buried after 1 seconds
worker 12 buried after 1 seconds
worker 13 buried after 1 seconds
worker 14 buried after 1 seconds
goodbye to uWSGI.
```

**Comment by xrmx** [comment]
Can you use a mule as thread pool instead of messing with workers? See http://uwsgi-docs.readthedocs.org/en/latest/Mules.html
Please always report bugs using latest stable release, 2.0.6 is quite an old one.

**Comment by mafanasev** [comment]
I'm using `MongoReplicaSetClient` in my app. In master i initialize connections to replica set to perform some database checking and after fork those connections become a part of workers. So I think that mule is not an option for me.
The problem is that every replica set client creates thread to monitor stepdowns, elections, and reconfigs  [(see this)](https://github.com/mongodb/mongo-python-driver/blob/master/pymongo/mongo_client.py#L388). And these sometimes leads to the deadlock.

I checked the bug with the latest version.
```
from flask import Flask
from pymongo import MongoReplicaSetClient

app = Flask(__name__)

MongoReplicaSetClient('localhost:27017', replicaset='foo')
MongoReplicaSetClient('localhost:27018', replicaset='foo')
MongoReplicaSetClient('localhost:27019', replicaset='foo')

@app.route('/')
def index():
    return 'OK'
```

```
$ pip freeze 
Flask==0.10.1
itsdangerous==0.24
Jinja2==2.8
MarkupSafe==0.23
pymongo==2.6.2
uWSGI==2.0.12
Werkzeug==0.11.3
wheel==0.24.0
```

```
$ for p in $(pgrep -f uwsgi); do gdb --batch -p $p -x bt; done | grep sem_wait
0x00007fde9556cfd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
0x00007fde9556cfd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
0x00007fde9556cfd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
```

```
*** Starting uWSGI 2.0.12 (64bit) on [Thu Jan  7 21:43:28 2016] ***
compiled with version: 4.6.3 on 07 January 2016 21:33:23
os: Linux-3.2.0-68-virtual #102-Ubuntu SMP Tue Aug 12 22:14:39 UTC 2014
nodename: localhost
machine: x86_64
clock source: unix
pcre jit disabled
detected number of CPU cores: 1
current working directory: /home/user/app
detected binary path: /home/user/app/env/bin/uwsgi
your processes number limit is 15964
your memory page size is 4096 bytes
detected max file descriptor number: 1024
lock engine: pthread robust mutexes
thunder lock: disabled (you can enable it with --thunder-lock)
uwsgi socket 0 bound to TCP address 0.0.0.0:8080 fd 4
Python version: 2.7.3 (default, Jun 22 2015, 19:44:33)  [GCC 4.6.3]
Python main interpreter initialized at 0x15d4360
python threads support enabled
your server socket listen backlog is limited to 100 connections
your mercy for graceful operations on workers is 60 seconds
mapped 3711168 bytes (3624 KB) for 50 cores
*** Operational MODE: preforking ***
WSGI app 0 (mountpoint='') ready in 0 seconds on interpreter 0x15d4360 pid: 26815 (default app)
*** uWSGI is running in multiple interpreter mode ***
spawned uWSGI master process (pid: 26815)
spawned uWSGI worker 1 (pid: 26832, cores: 1)
spawned uWSGI worker 2 (pid: 26833, cores: 1)
spawned uWSGI worker 3 (pid: 26834, cores: 1)
spawned uWSGI worker 4 (pid: 26835, cores: 1)
spawned uWSGI worker 5 (pid: 26836, cores: 1)
spawned uWSGI worker 6 (pid: 26837, cores: 1)
spawned uWSGI worker 7 (pid: 26838, cores: 1)
spawned uWSGI worker 8 (pid: 26839, cores: 1)
spawned uWSGI worker 9 (pid: 26840, cores: 1)
spawned uWSGI worker 10 (pid: 26841, cores: 1)
spawned uWSGI worker 11 (pid: 26842, cores: 1)
spawned uWSGI worker 12 (pid: 26843, cores: 1)
spawned uWSGI worker 13 (pid: 26844, cores: 1)
spawned uWSGI worker 14 (pid: 26845, cores: 1)
spawned uWSGI worker 15 (pid: 26846, cores: 1)
spawned uWSGI worker 16 (pid: 26847, cores: 1)
spawned uWSGI worker 17 (pid: 26848, cores: 1)
spawned uWSGI worker 18 (pid: 26849, cores: 1)
spawned uWSGI worker 19 (pid: 26850, cores: 1)
spawned uWSGI worker 20 (pid: 26851, cores: 1)
spawned uWSGI worker 21 (pid: 26852, cores: 1)
spawned uWSGI worker 22 (pid: 26853, cores: 1)
spawned uWSGI worker 23 (pid: 26854, cores: 1)
spawned uWSGI worker 24 (pid: 26855, cores: 1)
spawned uWSGI worker 25 (pid: 26856, cores: 1)
spawned uWSGI worker 25 (pid: 26856, cores: 1)
spawned uWSGI worker 26 (pid: 26857, cores: 1)
spawned uWSGI worker 27 (pid: 26858, cores: 1)
spawned uWSGI worker 28 (pid: 26859, cores: 1)
spawned uWSGI worker 29 (pid: 26860, cores: 1)
spawned uWSGI worker 30 (pid: 26861, cores: 1)
spawned uWSGI worker 31 (pid: 26862, cores: 1)
spawned uWSGI worker 32 (pid: 26863, cores: 1)
spawned uWSGI worker 33 (pid: 26864, cores: 1)
spawned uWSGI worker 34 (pid: 26869, cores: 1)
spawned uWSGI worker 35 (pid: 26870, cores: 1)
spawned uWSGI worker 36 (pid: 26871, cores: 1)
spawned uWSGI worker 37 (pid: 26872, cores: 1)
spawned uWSGI worker 38 (pid: 26873, cores: 1)
spawned uWSGI worker 39 (pid: 26874, cores: 1)
spawned uWSGI worker 40 (pid: 26875, cores: 1)
spawned uWSGI worker 41 (pid: 26876, cores: 1)
spawned uWSGI worker 42 (pid: 26877, cores: 1)
spawned uWSGI worker 43 (pid: 26878, cores: 1)
spawned uWSGI worker 44 (pid: 26879, cores: 1)
spawned uWSGI worker 45 (pid: 26880, cores: 1)
spawned uWSGI worker 46 (pid: 26881, cores: 1)
spawned uWSGI worker 47 (pid: 26882, cores: 1)
spawned uWSGI worker 48 (pid: 26883, cores: 1)
spawned uWSGI worker 49 (pid: 26884, cores: 1)
spawned uWSGI worker 50 (pid: 26885, cores: 1)
SIGINT/SIGQUIT received...killing workers...
worker 1 buried after 1 seconds
worker 2 buried after 1 seconds
worker 3 buried after 1 seconds
worker 4 buried after 1 seconds
worker 5 buried after 1 seconds
worker 6 buried after 1 seconds
worker 7 buried after 1 seconds
worker 8 buried after 1 seconds
worker 9 buried after 1 seconds
worker 10 buried after 1 seconds
worker 11 buried after 1 seconds
worker 12 buried after 1 seconds
worker 13 buried after 1 seconds
worker 17 buried after 1 seconds
worker 18 buried after 1 seconds
worker 19 buried after 1 seconds
worker 20 buried after 1 seconds
worker 21 buried after 1 seconds
worker 22 buried after 1 seconds
worker 23 buried after 1 seconds
worker 24 buried after 1 seconds
worker 25 buried after 1 seconds
worker 26 buried after 1 seconds
worker 27 buried after 1 seconds
worker 28 buried after 1 seconds
worker 29 buried after 1 seconds
worker 30 buried after 1 seconds
worker 31 buried after 1 seconds
worker 32 buried after 1 seconds
worker 33 buried after 1 seconds
worker 34 buried after 1 seconds
worker 35 buried after 1 seconds
worker 36 buried after 1 seconds
worker 37 buried after 1 seconds
worker 38 buried after 1 seconds
worker 39 buried after 1 seconds
worker 40 buried after 1 seconds
worker 41 buried after 1 seconds
worker 42 buried after 1 seconds
worker 43 buried after 1 seconds
worker 44 buried after 1 seconds
worker 45 buried after 1 seconds
worker 46 buried after 1 seconds
worker 47 buried after 1 seconds
worker 48 buried after 1 seconds
worker 49 buried after 1 seconds
worker 50 buried after 1 seconds
Thu Jan  7 21:52:00 2016 - worker 14 (pid: 26845) is taking too much time to die...NO MERCY !!!
Thu Jan  7 21:52:00 2016 - worker 15 (pid: 26846) is taking too much time to die...NO MERCY !!!
Thu Jan  7 21:52:00 2016 - worker 16 (pid: 26847) is taking too much time to die...NO MERCY !!!
worker 14 buried after 1 seconds
worker 15 buried after 1 seconds
worker 16 buried after 1 seconds
goodbye to uWSGI.
```

**Comment by unbit** [comment]
How do you re-spawn the mongo-monitoring thread after the master calls fork() ? from your description it looks like you ae expecting a thread to be inherited after fork

**Comment by mafanasev** [comment]
Yes, i expect that. Pymongo is able to recognize that fork was occured and then restores it's monitors
https://github.com/mongodb/mongo-python-driver/blob/master/pymongo/periodic_executor.py#L50

**Comment by unbit** [comment]
unless i am missing something this is not going to work, for python the "inherited" thread is running albeit for the kernel it is not a task so it will be never scheduled

**Comment by mafanasev** [comment]
Forget about "inherit". `MongoClient` instance has a reference for thread and it creates **new one** if thread is not started or is not alive. Thus it would be scheduled as usual.

**Comment by unbit** [comment]
but you said that the thread is spawned in the master, when the master calls fork() the worker will think the thread is running because it gets an exact copy of the address space.

**Comment by mafanasev** [comment]
No, `py-call-osafterfork` does the trick. It calls `PyOsAfterFork` whish calls [threading._after_fork](https://github.com/python/cpython/blob/master/Lib/threading.py#L1317) which resets the internal states of threads. So worker will think that thread is dead. I have already checked that with and without `py-call-osafterfork`.

**Comment by unbit** [comment]
have you already tried with --lazy-apps ? just to be sure it is really a forking problem

**Comment by mafanasev** [comment]
No i didn't. 
But i did the following: in master i stopped all the threads spawned by pymongo to be sure that only main thread is running at the fork() time. And it worked.

**Comment by unbit** [comment]
ok, so basically (as you said at the beginning) the problem is that the fixup hook assumes no python thread is running (so no GIL is acquired), but in your case the GIL is acquired (but cannot be released). It is an interesting problem, and i am not sure it is easy solvable as we cannot forcibly re-acquire the GIL without making mess

**Comment by mafanasev** [comment]
I also noticed that `uwsgi_python_master_fixup` is called before `PyOs_AfterFork`. We can rely on the GIL state only after `PyOs_AfterFork` because new GIL is created there. May be a wrong order is the problem.

**Comment by mihai-tiriplica-hs** [comment]
I have the same problem with uwsgi 2.0.11.2.

The backtrace for the hanging workers looks the same:

```
#0  0x00007fe0121cbfd0 in sem_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007fe010d9de28 in PyThread_acquire_lock () from /usr/lib/libpython2.7.so.1.0
#2  0x0000000000483029 in gil_real_get ()
#3  0x000000000047bb18 in uwsgi_python_master_fixup ()
#4  0x0000000000433ebe in uwsgi_respawn_worker ()
#5  0x00000000004326e9 in master_loop ()
#6  0x00000000004704c9 in uwsgi_run ()
#7  0x000000000041e30e in main ()
```

My config is: 

```
uwsgi --http-socket=:8080 --chdir=/opt/api --master --module=start --max-requests=1500 --harakiri=60 --processes=3 --enable-threads --ignore-sigpipe --die-on-term --worker-reload-mercy=5 --pidfile=/tmp/uwsgi.pid
```

Harakiri doesn't seem to work and the whole app hangs. I'm using MongoClient too.

**Comment by kevinxucs** [comment]
Seeing a similar issue with worker being hung with uWSGI 2.0.13.1

Backtrace from gdb:

```
#0  sem_wait () at ../nptl/sysdeps/unix/sysv/linux/x86_64/sem_wait.S:85
#1  0x00007f8be8fe5b78 in PyThread_acquire_lock () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
#2  0x00007f8be93dcf69 in gil_real_get () from /usr/local/lib/uwsgi/python27_plugin.so
#3  0x00007f8be93da189 in uwsgi_request_wsgi () from /usr/local/lib/uwsgi/python27_plugin.so
#4  0x0000000000418e8e in wsgi_req_recv ()
#5  0x000000000045ff14 in simple_loop_run ()
#6  0x00000000004640c4 in uwsgi_ignition ()
#7  0x00000000004687cd in uwsgi_worker_run ()
#8  0x0000000000468dbf in uwsgi_run ()
#9  0x000000000041855e in main ()
```

**Comment by poojaghumre** [comment]
Saw this same issue when using uwsgi 2.0.15 with openstack ceilometer api and the single child worker went into deadlock:
(gdb) thread apply all bt

Thread 1 (Thread 0x7f3d9555f840 (LWP 23989)):
#0  0x00007f3d95133a0b in do_futex_wait.constprop.1 () from /usr/lib64/libpthread.so.0
#1  0x00007f3d95133a9f in __new_sem_wait_slow.constprop.0 () from /usr/lib64/libpthread.so.0
#2  0x00007f3d95133b3b in sem_wait@@GLIBC_2.2.5 () from /usr/lib64/libpthread.so.0
#3  0x00007f3d9387f5f5 in PyThread_acquire_lock () from /usr/lib64/libpython2.7.so.1.0
#4  0x0000000000486649 in gil_real_get ()
#5  0x000000000047f2b8 in uwsgi_python_master_fixup ()
#6  0x0000000000434236 in uwsgi_respawn_worker ()
#7  0x000000000046c25c in uwsgi_start ()
#8  0x000000000046e321 in uwsgi_setup ()
#9  0x000000000041edf9 in main ()
(gdb) quit

**Comment by majorgreys** [comment]
@mafanasev @poojaghumre @kevinxucs #2388 introduced the `py-call-uwsgi-fork-hooks` option to fix the state of the internal interpreter before and after forking. It has been backported to the `uwsgi-2.0` branch. Would you be able to give it a try to confirm the resolves your deadlocks?

**Comment by pb-dod** [comment]
@majorgreys The upgrade to uwsgi 2.0.21 and enabling `py-call-uwsgi-fork-hooks` resolved the issue with deadlocks for me. I was seeing workers not come back up after a restart was trigged by `max-worker-lifetime`, and I could see "lifetime reached" as the last message in logs when I searched for "worker-1"/2/3/etc. Eventually all the workers would stop responding and it would fail liveness/readiness checks. The gdb output on a stuck worker process looked very similar to what @poojaghumre posted. I still don't know what first obtained the lock without releasing it.

Awesome work! Thanks so much for your fix. 

I'm wondering if the `harakiri` setting would have prevented this from getting so bad. As of uwsgi 2.0.21, I don't think `harakiri` works for killing individual long running threads, but maybe it would have killed a completely stuck worker in this case.

**Comment by poojaghumre** [comment]
Thank you for fixing this!
